### PR TITLE
impl(generator): format resource JSON as proto service

### PR DESCRIPTION
**Pull request by scotthart**
part of the work for #10980

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11154)

### CODE REVIEW

File: generator/internal/discovery_resource.h [hunk 1]
```diff
@@ -56,11 +64,14 @@ class DiscoveryResource {
   // are concatenated with the resource name.
   std::string FormatMethodName(std::string method_name) const;
 
+  StatusOr<std::string> JsonToProtoService() const;
```
Comment (dbolduc) [inline generator/internal/discovery_resource.h:67]: given that there is no json argument to this function, should we call it: `ToProtoService()` ?

File: generator/internal/discovery_resource.cc [hunk 2]
```diff
@@ -149,6 +176,63 @@ std::string DiscoveryResource::FormatMethodName(std::string method_name) const {
...
+    if (!rpc_options) return std::move(rpc_options).status();
+    rpc_text.push_back(*rpc_options);
+    rpc_text.emplace_back("  }");
+    rpcs_text.push_back(absl::StrJoin(rpc_text, "\n"));
+  }
+
+  return absl::StrCat(
+      FormatCommentBlock(absl::StrFormat(kServiceComments, name_, name_), 0),
+      absl::StrJoin(service_text, "\n"), absl::StrJoin(rpcs_text, "\n\n"),
+      "\n}\n\n");
```
Comment (dbolduc) [inline generator/internal/discovery_resource.cc:233]: Do we want the last service in a file to have these extra lines?

I suspect no, and it should be the caller's responsibility to add newlines between services.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11154)* status: 0 of 3 files reviewed, 5 unresolved discussions (waiting on @dbolduc)

___
*[`generator/internal/discovery_resource.h` line 36 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11154#-NRu6A11dUe6JaD8ctyu-r1-36:-NS7ITzkEiXqBifCN6RK:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/1ba233eaae9a321aab7b2e2d08afbd420e8f7f79/generator/internal/discovery_resource.h#L36)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

s/the path/path/
</blockquote></details>

Done.
___
*[`generator/internal/discovery_resource.h` line 67 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11154#-NRu6A11dUe6JaD8ctyu-r1-67:-NS7IUTQ5jjmRtbxlXXo:b-a5tpwk) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/1ba233eaae9a321aab7b2e2d08afbd420e8f7f79/generator/internal/discovery_resource.h#L67)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

given that there is no json argument to this function, should we call it: `ToProtoService()` ?
</blockquote></details>

The intent was to have the method name be the same as DiscoveryTypeVertex::JsonToProtobufMessage (adjusted it to match). I'm not particular what color these sheds are as long as they match, so if we want to change them I would do it in a subsequent PR.
___
*[`generator/internal/discovery_resource.cc` line 44 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11154#-NRu6A10bZgrFMBatbx6-r1-44:-NS7JEftBTpt05nql2Mm:bp3n5cb) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/1ba233eaae9a321aab7b2e2d08afbd420e8f7f79/generator/internal/discovery_resource.cc#L44)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

I don't think the first iteration needs to be handled outside of the loop. Consider:

```cc
  std::string output;
  std::size_t pos = 0;
  for (;;) {
    auto open = path.find('{', pos);
    if (open == std::string::npos) break;
    absl::StrAppend(&output, path.substr(pos, open + 1 - pos));
    pos = path.find('}', open);
    absl::StrAppend(
        &output, CamelCaseToSnakeCase(path.substr(open + 1, pos - open - 1)));
  }
  absl::StrAppend(&output, path.substr(pos));
  return output;
```
</blockquote></details>

I would prefer not to present the reader with an infinite loop with the exit condition(or conditions) somewhere in the body of the loop. I feel it's more expressive to keep the loop exit condition as part of the loop definition.
___
*[`generator/internal/discovery_resource.cc` line 212 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11154#-NRu6A10bZgrFMBatbx6-r1-212:-NS7Jo5s1LqZtP8B_OPh:bfvmsj2) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/1ba233eaae9a321aab7b2e2d08afbd420e8f7f79/generator/internal/discovery_resource.cc#L212)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

should `compute` be variable?
</blockquote></details>

Yes. Changed.
___
*[`generator/internal/discovery_resource.cc` line 225 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11154#-NRu6A10bZgrFMBatbx6-r1-225:-NS7JsnRCwOn9QxEwbrb:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/1ba233eaae9a321aab7b2e2d08afbd420e8f7f79/generator/internal/discovery_resource.cc#L225)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

nit: `*std::move(rpc_options)`
</blockquote></details>

Done.
___
*[`generator/internal/discovery_resource.cc` line 233 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11154#-NRu6A10bZgrFMBatbx6-r1-233:-NS7JtmJ6uYIImY58lvb:b-3mdqnl) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/1ba233eaae9a321aab7b2e2d08afbd420e8f7f79/generator/internal/discovery_resource.cc#L233)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

Do we want the last service in a file to have these extra lines?

I suspect no, and it should be the caller's responsibility to add newlines between services.
</blockquote></details>

Moved adding the extra line to the caller.

**Comment by coryan** [comment]
*[`generator/internal/discovery_resource.cc` line 44 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11154#-NRu6A10bZgrFMBatbx6-r1-44:-NS7k8Wu9mpFPAZ-AloA:b-yfk2gf) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/1ba233eaae9a321aab7b2e2d08afbd420e8f7f79/generator/internal/discovery_resource.cc#L44)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

I like:

```
for (;;) {
  i = Next();
  if (i == end) break;
  Foo();
}
```

You like:

```
i = Next();
while (i != end) {
  Foo();
  i = Next();
}
```

tomato tomato.
</blockquote></details>

```cc
for (i = Next(); i != end; i = Next()) {
  Foo();
}
```

**Review by dbolduc — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11154)* status: 1 of 3 files reviewed, 3 unresolved discussions (waiting on @coryan and @scotthart)

___
*[`generator/internal/discovery_resource.cc` line 44 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11154#-NRu6A10bZgrFMBatbx6-r1-44:-NS7l3fp65vUX74nWnKz:b-m1fq9q) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/1ba233eaae9a321aab7b2e2d08afbd420e8f7f79/generator/internal/discovery_resource.cc#L44)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

```cc
for (i = Next(); i != end; i = Next()) {
  Foo();
}
```

</blockquote></details>

My preference changes depending on how many characters it takes to type `Next()` :)

**Review by dbolduc — APPROVED** [review]

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11154)* status: 1 of 3 files reviewed, 3 unresolved discussions (waiting on @coryan and @dbolduc)

___
*[`generator/internal/discovery_resource.h` line 67 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11154#-NRu6A11dUe6JaD8ctyu-r1-67:-NS7l83I0bKYzma3byq5:bc7i014) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/1ba233eaae9a321aab7b2e2d08afbd420e8f7f79/generator/internal/discovery_resource.h#L67)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

Ack. I would drop the `Json-` prefix. But it's a nit.
</blockquote></details>

As long as they are consistent, I'm indifferent to whether `Json` is present in the name.
___
*[`generator/internal/discovery_resource.cc` line 44 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11154#-NRu6A10bZgrFMBatbx6-r1-44:-NS7m5Vj-V6NB9kOXeNK:bwejg1r) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/1ba233eaae9a321aab7b2e2d08afbd420e8f7f79/generator/internal/discovery_resource.cc#L44)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

My preference changes depending on how many characters it takes to type `Next()` :)
</blockquote></details>

Already refactored the `while` to a `for`. 
I'm more interested in expressiveness than the code golf score.
___
*[`generator/internal/discovery_resource.cc` line 46 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11154#-NS7fToT9-B0juo8CzJa:-NS7m3BO4sAmbH1RpvSc:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/1ba233eaae9a321aab7b2e2d08afbd420e8f7f79/generator/internal/discovery_resource.cc#L46)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

Delete this line. It does not need special handling.
</blockquote></details>

Done.
___
*[`generator/internal/discovery_resource.cc` line 58 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11154#-NS7fcLnAIAZKr2u91ou:-NS7m2rD7MYAQew2LLy6:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/1ba233eaae9a321aab7b2e2d08afbd420e8f7f79/generator/internal/discovery_resource.cc#L58)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

`open == std::string::npos` here so we should just say `path.substr(current)`
</blockquote></details>

Done.
___
*[`generator/internal/discovery_resource.cc` line 212 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11154#-NRu6A10bZgrFMBatbx6-r1-212:-NS7m1HsDf2dq3TtKv3B:b-80tlqk) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/1ba233eaae9a321aab7b2e2d08afbd420e8f7f79/generator/internal/discovery_resource.cc#L212)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

Ahh, then I am assuming we should also `s/v1/$version$/` ?
</blockquote></details>

Sure.

**Pull request merged**